### PR TITLE
Support for generating Xcode project from cmakefiles

### PR DESCRIPTION
**Pull request by Zorro666**
## Description

Support for the cmake Xcode project generator

Example command line to generate Xcode project instead of makefiles

`cmake -H. -Bbuild-xcode -G Xcode `

### Changes required to fix building from Xcode instead of makefiles
- Xcode does not support having the same source file with different extensions.
  - Renamed cgl_platform.mm -> cgl_platform_helpers.mm
  - Renamed vk_apple.mm -> vk_apple_helpers.mm
- Use MAKE generator for qrenderdoc custom_commands
- For Xcode set the output directory to be the same for all configurations
- Building via Xcode does not include "stdc++" by default (probably a difference between cland and clang++).
- Pass the full path to the static library "renderdoc_libentry" on the link line.
- Capture the PATH variable at cmake invoke time and store in the cache.
- Use the captured PATH variable for the make based build targets custom_swig and QRenderDoc.

### Tweaks to make the generated Xcode project nicer to work with
- Hardcoded addition for qrenderdoc sources and headers to the Xcode project currently set to `*.c*, *.h*, *.m*` in the `Code`, `Styles`, `Widgets` and `Windows` source folders.
- Restrict the schemes to match the configuration on the cmake command line ie. Debug/Release
- Disable Xcode Debug Document Versioning
- Specify the executable path for the build-qrenderdoc target

### Minor tidy ups in the Apple-specific parts of the existing cmakefile
- Apple qrenderdoc cmake remove extra space in RPATH
- Make QMAKE_RPATHDIR Apple-specific
- Make OSX_ICONFILE Apple-specific
- Make QMAKE_OSX_DEPLOYMENT_TARGET Apple-specific

Demonstration of project building inside Xcode

Screenshot 2021-06-04 at 22 33 31

Demonstration of Xcode semantic analysis detecting compile errors without having to build 

image

Demonstration of autocomplete working

image

### CODE REVIEW

File: qrenderdoc/CMakeLists.txt [hunk 1]
```diff
@@ -185,8 +187,15 @@ file(WRITE
      "LIBS+=${PYTHON_LINK}\n"
      "\n"
      "RENDERDOC_VERSION=${RENDERDOC_VERSION}\n"
-     "\n"
-     "OSX_ICONFILE=${CMAKE_CURRENT_BINARY_DIR}/RenderDoc.icns\n")
+     "\n")
+
+if(APPLE)
```
Comment (baldurk) [inline qrenderdoc/CMakeLists.txt:195]: I don't understand why this code has been moved out into this `if()`? `QMAKE_RPATHDIR` is the only property that would be relevant for other platforms and it's not set to anything except on apple.

File: qrenderdoc/CMakeLists.txt [hunk 2]
```diff
@@ -309,11 +318,20 @@ if(ENABLE_QRENDERDOC)
...
 # can do dependency checking and rebuild anything necessary.
 add_custom_command(OUTPUT QRenderDoc
-    COMMAND ${QMAKE_QT5_COMMAND} "CMAKE_DIR=${CMAKE_BINARY_DIR}" ${CMAKE_CURRENT_SOURCE_DIR}
+    COMMAND ${SET_SYSTEM_PATH_COMMAND} ${QMAKE_QT5_COMMAND} "CMAKE_DIR=${CMAKE_BINARY_DIR}" ${CMAKE_CURRENT_SOURCE_DIR}
     COMMAND ${GENERATOR_MAKE} ${GENERATOR_MAKE_PARAMS}
     DEPENDS RenderDoc.icns)
 add_custom_target(build-qrenderdoc ALL DEPENDS QRenderDoc DEPENDS renderdoc DEPENDS swig-bindings)
 
+if(APPLE)
+    set_target_properties(build-qrenderdoc PROPERTIES XCODE_GENERATE_SCHEME ON)
```
Comment (baldurk) [inline qrenderdoc/CMakeLists.txt]: Is this line necessary? The documentation for `CMAKE_XCODE_GENERATE_SCHEME` suggests that it propagates to `XCODE_GENERATE_SCHEME` on all targets.

**Review by baldurk — APPROVED** [review]
I think this looks fine overall and for the most part only affects apple/xcode anyway. I've left a couple of queries on some changes that I'm not sure about but they don't necessarily need to change as long as there's a reason to keep them as-is.

File: qrenderdoc/CMakeLists.txt (hunk 2 above)
Comment (Zorro666) [reply qrenderdoc/CMakeLists.txt]: I do not recall why this was here - perhaps this was done first.
I have removed it locally and the Xcode project seems to be behaving the same.

File: qrenderdoc/CMakeLists.txt (hunk 1 above)
Comment (Zorro666) [reply qrenderdoc/CMakeLists.txt:195]: No particular reason tidying up the APPLE specific changes to be behind if (APPLE).
When QMAKE_RPATHDIR was added did not notice the pattern to use if (APPLE) and append to the qmake config file. 

Happy to revert this change

File: qrenderdoc/CMakeLists.txt (hunk 2 above)
Comment (Zorro666) [reply qrenderdoc/CMakeLists.txt]: Happy to remove this line

File: qrenderdoc/CMakeLists.txt (hunk 1 above)
Comment (Zorro666) [reply qrenderdoc/CMakeLists.txt:195]: Dropped the commit from history

File: qrenderdoc/CMakeLists.txt (hunk 2 above)
Comment (Zorro666) [reply qrenderdoc/CMakeLists.txt]: Reverted line and squashed into

https://github.com/baldurk/renderdoc/pull/2287/commits/92745c6436606e0df109fcb221ab58a8eedacf7d

**Pull request merged**